codec-stomp/src/main/java/io/netty/handler/codec/stomp/StompSubframeDecoder.java

```diff
@@ -266,28 +259,71 @@ private static void skipControlCharacters(ByteBuf buffer) {
         }
     }
 
-    private static String readLine(ByteBuf buffer, int maxLineLength) {
-        AppendableCharSequence buf = new AppendableCharSequence(128);
+    private String readLine(ByteBuf buffer, int initialBufferSize) {
+        AppendableCharSequence buf = new AppendableCharSequence(initialBufferSize);
         int lineLength = 0;
         for (;;) {
             byte nextByte = buffer.readByte();
             if (nextByte == StompConstants.CR) {
-                nextByte = buffer.readByte();
-                if (nextByte == StompConstants.LF) {
-                    return buf.toString();
-                }
+                //do nothing
             } else if (nextByte == StompConstants.LF) {
                 return buf.toString();
             } else {
                 if (lineLength >= maxLineLength) {
-                    throw new TooLongFrameException("An STOMP line is larger than " + maxLineLength + " bytes.");
+                    invalidLineLength();
                 }
                 lineLength ++;
                 buf.append((char) nextByte);
             }
         }
     }
 
+    private boolean readHeader(StompHeaders headers, AppendableCharSequence buf, ByteBuf buffer) {
+        buf.reset();
+        int lineLength = 0;
+        String key = null;
+        boolean valid = false;
+        for (;;) {
+            byte nextByte = buffer.readByte();
+
+            if (nextByte == StompConstants.COLON && key == null) {
+                key = buf.toString();
+                valid = true;
+                buf.reset();
+            } else if (nextByte == StompConstants.CR) {
+                //do nothing
+            } else if (nextByte == StompConstants.LF) {
+                if (key == null && lineLength == 0) {
+                    return false;
+                } else if (valid) {
+                    headers.add(key, buf.toString());
+                } else if (validateHeaders) {
+                    invalidHeader(key, buf.toString());
+                }
+                return true;
+            } else {
+                if (lineLength >= maxLineLength) {
+                    invalidLineLength();
+                }
+                if (nextByte == StompConstants.COLON && key != null) {
+                    valid = false;
```

Comment:
Maybe we should call an `invalidHeader()` here?

Reply:
While the code would certainly look better, it would have side effects:
- the generated error message would not contain the entire header
- in case validation is disabled, the next header would be read starting after this point, not after the line break

Reply:
Ok :)